### ISSUE TITLE
MTPLossAutoScaler loss scale bugs when cp enabled

### ISSUE THREAD

**Issue by Baibaifan** [NVIDIA/Megatron-LM#3943]
**Describe the bug**

> 1. Gradient scale in DDP with DP size. https://github.com/NVIDIA/Megatron-LM/blob/9ed8b0c4ac8e65d80fb35eecb3656c96b30df943/megatron/core/distributed/distributed_data_parallel.py#L291
> 2. When enabled cp, `num_microbatches` here will be multiplied by cp. https://github.com/NVIDIA/Megatron-LM/blob/9ed8b0c4ac8e65d80fb35eecb3656c96b30df943/megatron/core/pipeline_parallel/schedules.py#L311

The loss was reduced by a factor of cp.

reference：https://github.com/NVIDIA/Megatron-LM/pull/2217

**Steps/Code to reproduce bug**

```bash
--context-parallel-size > 1
--mtp-num-layers > 1
--mtp-loss-scaling-factor > 0.0
```

**Comment by BestJuly** [comment]
Hi @Baibaifan , after careful analysis, I believe MTP loss does **not** have the same scaling issue as MoE auxiliary loss under CP, and no fix is needed.

### Root cause of the difference

The MoE aux loss fix ([PR #2217](https://github.com/NVIDIA/Megatron-LM/pull/2217)) is needed because `MoEAuxLossAutoScaler` receives a **partial sum** — [`probs.sum(dim=0)`](https://github.com/NVIDIA/Megatron-LM/blob/768855703fc78b96302f788697e55b698b4591e1/megatron/core/transformer/moe/moe_utils.py#L138) sums over local tokens only, making `aux_loss ∝ 1/cp_size`. After DDP averaging across `dp_cp_group`, the gradient is reduced by an extra factor of `cp_size`, requiring compensation.

MTP loss does **not** have this problem because, even when `calculate_per_token_loss=False`, [`process_mtp_loss`](https://github.com/NVIDIA/Megatron-LM/blob/768855703fc78b96302f788697e55b698b4591e1/megatron/core/transformer/multi_token_prediction.py#L703-L707) already computes a **per-token mean** before passing it to `MTPLossAutoScaler`:

```python
safe_num_tokens = num_tokens.clamp(min=1)
hidden_states = MTPLossAutoScaler.apply(
    hidden_states, mtp_loss_scale * mtp_loss / safe_num_tokens  # per-token mean
)
```

If try to fix this using the similar strategy as [PR#2217](https://github.com/NVIDIA/Megatron-LM/pull/2217), I observed loss divergence for the lm_loss and mtp loss.

### Follow-up
Not sure whether you have observed the issue for a specific config so my test may not cover. If it is the case, could you please provide a repro so I can investigate further? Thank you.

**Comment by Baibaifan** [comment]
> Hi [@Baibaifan](https://github.com/Baibaifan) , after careful analysis, I believe MTP loss does **not** have the same scaling issue as MoE auxiliary loss under CP, and no fix is needed.
> 
> ### Root cause of the difference
> The MoE aux loss fix ([PR #2217](https://github.com/NVIDIA/Megatron-LM/pull/2217)) is needed because `MoEAuxLossAutoScaler` receives a **partial sum** — [`probs.sum(dim=0)`](https://github.com/NVIDIA/Megatron-LM/blob/768855703fc78b96302f788697e55b698b4591e1/megatron/core/transformer/moe/moe_utils.py#L138) sums over local tokens only, making `aux_loss ∝ 1/cp_size`. After DDP averaging across `dp_cp_group`, the gradient is reduced by an extra factor of `cp_size`, requiring compensation.
> 
> MTP loss does **not** have this problem because, even when `calculate_per_token_loss=False`, [`process_mtp_loss`](https://github.com/NVIDIA/Megatron-LM/blob/768855703fc78b96302f788697e55b698b4591e1/megatron/core/transformer/multi_token_prediction.py#L703-L707) already computes a **per-token mean** before passing it to `MTPLossAutoScaler`:
> 
> safe_num_tokens = num_tokens.clamp(min=1)
> hidden_states = MTPLossAutoScaler.apply(
>     hidden_states, mtp_loss_scale * mtp_loss / safe_num_tokens  # per-token mean
> )
> If try to fix this using the similar strategy as [PR#2217](https://github.com/NVIDIA/Megatron-LM/pull/2217), I observed loss divergence for the lm_loss and mtp loss.
> 
> ### Follow-up
> Not sure whether you have observed the issue for a specific config so my test may not cover. If it is the case, could you please provide a repro so I can investigate further? Thank you.

hi, @BestJuly 

https://github.com/NVIDIA/Megatron-LM/blob/9a60a18ef731214be3752d8839ad07b9ac2b4111/megatron/core/transformer/multi_token_prediction.py#L598

https://github.com/NVIDIA/Megatron-LM/blob/9a60a18ef731214be3752d8839ad07b9ac2b4111/megatron/core/transformer/multi_token_prediction.py#L706


Forward calculation: `scaled_mtp_loss = mtp_loss_scale * mtp_loss / safe_num_tokens`; According to the chain rule, the backward  calculation is:

<img width="1002" height="376" alt="Image" src="https://github.com/user-attachments/assets/846d4372-3b9a-4f10-8425-f1b9d7d60fcf" />

```python
grad_mtp_loss =
    mtp_loss_backward_scale
    * (mtp_loss_scale / safe_num_tokens)
```
	​
Am I right in my understanding?

**Comment by BestJuly** [comment]
Hi @Baibaifan , I think your chain-rule derivation for a single rank is correct, but I think this is still only the local CP-shard view.

After backward, gradients are synchronized over the DP+CP group, and DDP scaling is done using [`dp_cp_group.size()`](https://github.com/NVIDIA/Megatron-LM/blob/9ed8b0c4ac8e65d80fb35eecb3656c96b30df943/megatron/core/distributed/distributed_data_parallel.py#L288-L291), i.e. CP ranks are already included in the intended averaging.

I think this is the key difference from MoE aux loss:
- MoE aux loss enters the autoscaler as a **partial sum over local tokens**, so CP averaging causes an extra shrink and needs compensation.
- MTP enters the autoscaler as a **local per-token mean** (already divided by `safe_num_tokens`), so the later averaging across `dp_cp_group` is part of the intended normalization, not an extra bug.

So I think your formula is correct as a **local derivative expression**, but it should not be interpreted as evidence that MTP is missing an additional `cp_size` factor.

Also, by saying divergence, I also validated with a similar code change with [PR2217](https://github.com/NVIDIA/Megatron-LM/pull/2217), which is [PR3959](https://github.com/NVIDIA/Megatron-LM/pull/3959). The CP2 loss is more close to CP1 for original code (current main branch) instead of adding the fix.

<img width="2854" height="626" alt="Image" src="https://github.com/user-attachments/assets/03bb9e49-277c-454b-a81d-6b900ba1d7dd" />

**Comment by Baibaifan** [comment]
> Hi [@Baibaifan](https://github.com/Baibaifan) , I think your chain-rule derivation for a single rank is correct, but I think this is still only the local CP-shard view.
> 
> After backward, gradients are synchronized over the DP+CP group, and DDP scaling is done using [`dp_cp_group.size()`](https://github.com/NVIDIA/Megatron-LM/blob/9ed8b0c4ac8e65d80fb35eecb3656c96b30df943/megatron/core/distributed/distributed_data_parallel.py#L288-L291), i.e. CP ranks are already included in the intended averaging.
> 
> I think this is the key difference from MoE aux loss:
> 
> * MoE aux loss enters the autoscaler as a **partial sum over local tokens**, so CP averaging causes an extra shrink and needs compensation.
> * MTP enters the autoscaler as a **local per-token mean** (already divided by `safe_num_tokens`), so the later averaging across `dp_cp_group` is part of the intended normalization, not an extra bug.
> 
> So I think your formula is correct as a **local derivative expression**, but it should not be interpreted as evidence that MTP is missing an additional `cp_size` factor.
> 
> Also, by saying divergence, I also validated with a similar code change with [PR2217](https://github.com/NVIDIA/Megatron-LM/pull/2217), which is [PR3959](https://github.com/NVIDIA/Megatron-LM/pull/3959). The CP2 loss is more close to CP1 for original code (current main branch) instead of adding the fix.
> 
> <img alt="Image" width="2000" height="626" src="https://private-user-images.githubusercontent.com/19769279/567495550-03bb9e49-277c-454b-a81d-6b900ba1d7dd.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQyMzI0NzgsIm5iZiI6MTc3NDIzMjE3OCwicGF0aCI6Ii8xOTc2OTI3OS81Njc0OTU1NTAtMDNiYjllNDktMjc3Yy00NTRiLWE4MWQtNmI5MDBiYTFkN2RkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjAzMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwMzIzVDAyMTYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWM1MzRhNDcyOTdlOWMwZDQ2NDk3YjVkNzlmNDk0MGUxMWY5ZDQwNzkzNmJmNjZhYWY3YzZkYTQ3ODZlOTg2NTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.PJGsJUdPQycGdd-u0GspDtkn61N-2f2MdNkUZ0d5_U8">

@BestJuly 

I actually have two questions:
1. Is `mtp_loss` based on the local CP per_token logic? If so, is the `set_loss_scale` logic based on the sequence(`MTPLossAutoScaler.set_loss_scale(loss_scale / num_microbatches)`, num_microbatches = mini_batch / dp_size)?

2. Does `mtp_loss_scale * mtp_loss / safe_num_tokens` become `mtp_loss_backward_scale * (mtp_loss_scale / safe_num_tokens)` on the reverse side according to the chain rule? Is this reasonable, since `safe_num_tokens = local CP  token nums`?

From the formula's perspective, I find the idea of ​​`local CP per_token loss -> sequence-level `mtp_loss_backward_scale` rather strange?

**Comment by BestJuly** [comment]
Hi @Baibaifan  , I may make this complicated. When MTP part is built using MoE layers, one MTP layer should have two losses: MTP loss and MTP aux loss. 
- MTP loss
    - The per-token loss is for MTP loss part, similar to the lm loss but the code paths are different. 
    - It already uses local CP-shard per-token loss so there is no need to change the scale. 
    - In the [test_multi_token_prediction.py](https://github.com/NVIDIA/Megatron-LM/blob/9ed8b0c4ac8e65d80fb35eecb3656c96b30df943/tests/unit_tests/transformer/test_multi_token_prediction.py#L338) we have tests to make sure the loss can be aligned using different tp/cp combination with tp1cp1.
- MTP MoE aux loss: the fix in [PR2217](https://github.com/NVIDIA/Megatron-LM/pull/2217) should have also fixed this part.
